runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/JobService.java

```diff
@@ -0,0 +1,340 @@
+import org.apache.beam.sdk.fn.stream.SynchronizedStreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A JobService that prepares and runs jobs on behalf of a client using a {@link JobInvoker}.
+ *
+ * <p>Job management is handled in-memory rather than any persistent storage, running the risk of
+ * leaking jobs if the JobService crashes.
+ *
+ * <p>TODO: replace in-memory job management state with persistent solution.
+ */
+public class JobService extends JobServiceGrpc.JobServiceImplBase implements FnService {
+  private static final Logger LOG = LoggerFactory.getLogger(JobService.class);
+
+  public static JobService create(
+      ArtifactStagingServiceProvider artifactStagingServiceProvider, JobInvoker invoker) {
+    return new JobService(artifactStagingServiceProvider, invoker);
+  }
+
+  private final ConcurrentMap<String, JobPreparation> preparations;
+  private final ConcurrentMap<String, JobInvocation> invocations;
+  private final ArtifactStagingServiceProvider artifactStagingServiceProvider;
+  private final JobInvoker invoker;
+
+  private JobService(
+      ArtifactStagingServiceProvider artifactStagingServiceProvider, JobInvoker invoker) {
+    this.artifactStagingServiceProvider = artifactStagingServiceProvider;
+    this.invoker = invoker;
+
+    this.preparations = new ConcurrentHashMap<>();
+    this.invocations = new ConcurrentHashMap<>();
+  }
+
+  @Override
+  public void prepare(
+      PrepareJobRequest request,
+      StreamObserver<PrepareJobResponse> responseObserver) {
+    try {
+      LOG.trace("{} {}", PrepareJobRequest.class.getSimpleName(), request);
+      // insert preparation
+      String preparationId =
+          String.format("%s_%d", request.getJobName(), ThreadLocalRandom.current().nextInt());
+      GrpcFnServer<ArtifactStagingService> stagingService =
+          artifactStagingServiceProvider.forJob(preparationId);
+      Struct pipelineOptions = request.getPipelineOptions();
+      if (pipelineOptions == null) {
+        LOG.trace("PIPELINE OPTIONS IS NULL");
+        throw new NullPointerException("Encountered null pipeline options.");
+            /*
+        LOG.debug("Encountered null pipeline options.  Using default.");
+        pipelineOptions = Struct.getDefaultInstance();
+        */
+      } else {
+        LOG.trace("PIPELINE OPTIONS IS NOT NULL");
+      }
+      LOG.trace("PIPELINE OPTIONS {} {}", pipelineOptions.getClass(), pipelineOptions);
+      JobPreparation preparation =
+          JobPreparation
+              .builder()
+              .setId(preparationId)
+              .setPipeline(request.getPipeline())
+              .setOptions(pipelineOptions)
+              .setStagingService(stagingService)
+              .build();
+      JobPreparation previous = preparations.putIfAbsent(preparationId, preparation);
+      if (previous != null) {
+        // retry recursively in the unlikely case of a name collision.
```

Comment:
This is funky, and we are forgetting to close stagingService. I recommend to replace nextInt() with generating a UUID so there are no collisions, then this issue is avoided.